### PR TITLE
FIX: REVIEW AND CORRECT CLASS ENDPOINTS

### PR DESCRIPTION
**Pull request by ivamorgra**
# Proposed Changes in Class Endpoints

## Description of Changes
In this pull request, I've made changes and revisions to the endpoints related to classes. The key modification is the addition of a specific middleware that uses JSON Web Token (JWT) to authorize the user before granting access to certain resources.

## Details of Changes
- Added a JWT authorization middleware to each class endpoint.
- Changes include verification of the JWT token to ensure the user is authenticated before processing the request.

## Actions for Reviewer
Before approving the pull request, please follow this checklist to ensure the changes work correctly in Postman:

- [x] **Authentication:** Verify that JWT authorization is correctly implemented.
- [x] **Endpoints:** Run Postman requests for each modified endpoint.
- [x] **Valid Requests:** Ensure valid requests with a valid token are properly authorized.
- [x] **Invalid Requests:** Confirm that invalid requests without a token or with an invalid token are appropriately rejected.
- [x] **Error Handling:** Check that authorization errors are handled appropriately.

## Additional Notes
Please provide specific feedback on any issues you may identify during the review. Thank you.

### CODE REVIEW

**Comment by ivamorgra** [comment]
I apologize for the initial commit, I confused the task number.

File: server/routes/classes.ts [hunk 1]
```diff
@@ -12,6 +13,10 @@ const storage = new Storage({
 const bucketName = 'classes-bucket'
 const bucket = storage.bucket(bucketName)
 
+const ERROR_LOGIN = 'Unauthenticated: You are not logged in'
```
Comment (mrarjona) [inline server/routes/classes.ts]: I think you can remove this constant, it is not used in this file.

File: server/routes/classes.ts [hunk 2]
```diff
@@ -169,50 +170,53 @@ router.put(
...
-        } catch (error) {
-            return res
-                .status(500)
-                .json({ error: 'Error deleting file from bucket' })
-        }
 
-        await classData.deleteOne()
-        return res.status(200).json({ message: 'Class deleted successfully' })
+            await classData.deleteOne()
+            return res.status(200).json({ message: 'Class deleted successfully' })
```
Comment (mrarjona) [inline server/routes/classes.ts]: For the correct deletion code it was defined in the OAS that the status code should be 204. You can look at the material delete operation to put it similar.

File: server/routes/classes.ts [hunk 3]
```diff
@@ -169,50 +170,53 @@ router.put(
...
-            const savedClass = await _class.save()
-            return res.status(200).json(savedClass)
+        }catch{
+            return res.status(500).json({ error: ERROR_SERVER })
         }
     }
 )
 
-router.delete('/:id', async (req: Request, res: Response) => {
+router.delete('/:id',authUser, async (req: Request, res: Response) => {
```
Comment (mrarjona) [inline server/routes/classes.ts:194]: This operation must also take into account the status code 403.

File: server/routes/classes.ts [hunk 4]
```diff
@@ -74,63 +74,56 @@ router.post('/', upload.single('file'), async (req: Request, res: Response) => {
...
-        res.status(500).json({ error: 'Internal Server Error' })
+        res.status(500).json({ error: ERROR_SERVER })
     }
 })
 
-router.put(
-    '/:id',
-    upload.single('file'),
-    async (req: Request, res: Response) => {
+router.put('/:id',authUser,upload.single('file'),async (req: Request, res: Response) => {
```
Comment (mrarjona) [inline server/routes/classes.ts:108]: This operation must also take into account the status code 403.

File: server/routes/classes.ts [hunk 5]
```diff
@@ -169,50 +170,53 @@ router.put(
...
-            return res.status(200).json(savedClass)
+        }catch{
+            return res.status(500).json({ error: ERROR_SERVER })
         }
     }
 )
 
-router.delete('/:id', async (req: Request, res: Response) => {
+router.delete('/:id',authUser, async (req: Request, res: Response) => {
     //TODO: Check if user is the author of the class
```
Comment (mrarjona) [inline server/routes/classes.ts:195]: Delete this "TODO" when you are aware of status code 403.

File: server/routes/classes.ts [hunk 6]
```diff
@@ -74,63 +74,56 @@ router.post('/', upload.single('file'), async (req: Request, res: Response) => {
...
+        res.status(500).json({ error: ERROR_SERVER })
     }
 })
 
-router.put(
-    '/:id',
-    upload.single('file'),
-    async (req: Request, res: Response) => {
+router.put('/:id',authUser,upload.single('file'),async (req: Request, res: Response) => {
         //TODO: Check if user is the author of the class
```
Comment (mrarjona) [inline server/routes/classes.ts:109]: Delete this "TODO" when you are aware of status code 403.

File: server/routes/classes.ts [hunk 7]
```diff
@@ -26,29 +31,24 @@ function getFileNameFromUrl(url: string): string | null {
...
 
-this is responsible of another microservice
-
-*/
+})
 
 //TODO: Connect to course microservice in frontend
 
-router.post('/', upload.single('file'), async (req: Request, res: Response) => {
+router.post('/',authUser, upload.single('file'), async (req: Request, res: Response) => {
```
Comment (mrarjona) [inline server/routes/classes.ts]: With respect to the OAS defined, the endpoint associated to the POST must be "/classes/course/{courseId}" to know in which course the class is being uploaded.

**Review by mrarjona — CHANGES_REQUESTED** [review]
Good work @ivamorgra! I have commented some status codes that you are missing to check, and a change you should make on the endpoint defined for the POST operation.

If you can, move the task associated with this Pull Request in the project dashboard to In Review.

Make use in the app.ts file of a middleware to be able to parse multer errors due to large files (more than 1GB), you can take a look at how it is done with materials.

In case it is easier for you to fix the issue of status codes, endpoints, etc. I share with you a [new version of the OAS](https://app.swaggerhub.com/apis/mrarjona/learning-microservice/0.1.0#/) that I have just written and should be complete. Please let me know if you find any errors.

File: server/routes/classes.ts (hunk 1 above)
Comment (ivamorgra) [reply server/routes/classes.ts]: Yes!!! I changed the function to another file and I forgot this. Good point!

File: server/routes/classes.ts (hunk 2 above)
Comment (ivamorgra) [reply server/routes/classes.ts]: Ok!

**Comment by ivamorgra** [comment]
Thank you for the feedback @migromarj! Currently, I'm unable to implement the handling of 403 errors as it requires communication with the courses microservice, which is not yet in place. Once the necessary integration is completed, I'll ensure to incorporate proper handling for the 403 status code. Appreciate your understanding!

**Comment by mrarjona** [comment]
> Thank you for the feedback @migromarj! Currently, I'm unable to implement the handling of 403 errors as it requires communication with the courses microservice, which is not yet in place. Once the necessary integration is completed, I'll ensure to incorporate proper handling for the 403 status code. Appreciate your understanding!

You're absolutely right, I'll bear that in mind!

File: server/routes/classes.ts [hunk 8]
```diff
@@ -12,12 +13,16 @@ const storage = new Storage({
...
+const ERROR_SERVER = 'Internal Server Error'
+
 const allowedMimeTypes = ['video/mp4', 'video/mpeg', 'video/quicktime']
 
 const upload = multer({
     storage: multer.memoryStorage(),
     limits: {
-        fileSize: 1024 * 1024 * 1024, // 1 GB
+        fileSize: 1024 * 1024 * 5, // 1 GB
+        //fileSize: 1024 * 1024 * 1024, // 1 GB
```
Comment (mrarjona) [inline server/routes/classes.ts]: Uncomment this line and delete the previous one to set a limit of 1GB and not 5MB.

**Review by mrarjona — APPROVED** [review]
Thank you for taking the suggestions into account @ivamorgra. I have accepted it, everything works correctly, but **before making the merge change the maximum file space from 5MB to 1GB**, I have left in a comment above the modifications to be made.

**Pull request merged**